File: 54.spiral-matrix/solution.cpp
class Solution {
public:
	vector<int> spiralOrder(vector<vector<int>>& matrix) {
		vector<int> result;
		if (matrix.empty()) return result;
		int i = 0, j = -1, dir = 0;
		int goal;
		int left = 0;
		while (true)
		{
			switch (dir)
			{
			case 0:
				goal = matrix.front().size() - 1 - left;
				if (goal <= j) return result;
				for (j++; j <= goal; j++) result.push_back(matrix[i][j]);
				j--;
				break;
			case 1:
				goal = matrix.size() - 1 - left;
				if (goal <= i) return result;
				for (i++; i <= goal; i++) result.push_back(matrix[i][j]);
				i--;
				break;
			case 2:
				goal = left;
				if (goal >= j) return result;
				for (j--; j >= goal; j--) result.push_back(matrix[i][j]);
				j++;
				break;
			case 3:
				goal = ++left;
				if (goal >= i) return result;
				for (i--; i >= goal; i--) result.push_back(matrix[i][j]);
				i++;
			}
			dir = (dir + 1) % 4;
		}
		return result;
	}
};

File: 54.spiral-matrix/solution.go
func spiralOrder(matrix [][]int) []int {
	if len(matrix) == 0 {
		return []int{}
	}
	length := len(matrix) * len(matrix[0])
	result := make([]int, length, length)
	ii, jj := 0, 1
	matrixI, matrixJ := 0, 0
	dir := 1
	for i := 0; i < len(result); i++ {
		result[i] = matrix[matrixI][matrixJ]
		switch dir {
		case 1:
			if len(matrix[0])-matrixJ-matrixI == 1 {
				dir = 2
				ii, jj = jj, -ii
			}
		case 2:
			if len(matrix)-matrixI == len(matrix[0])-matrixJ {
				dir = 3
				ii, jj = jj, -ii
			}
		case 3:
			if len(matrix)-matrixI-matrixJ == 1 {
				dir = 4
				ii, jj = jj, -ii
			}
		case 4:
			if matrixI-matrixJ == 1 {
				dir = 1
				ii, jj = jj, -ii
			}
		}
		matrixI += ii
		matrixJ += jj
	}
	return result
}
